pkg/kube/krt/README.md

```diff
@@ -0,0 +1,157 @@
+These are *not* expressed as generic constraints due to limitations in Go's type system.
+
+* Each object `T` must have a unique `Key[T]` (which is just a typed wrapper around `string`) that uniquely identifies the object.
+    Default implementations exist for Kubernetes objects, Istio `config.Config` objects, and `ResourceName() string` implementations.
+* `Equals(k K) bool` may be implemented to provide custom implementations to compare objects. Comparison is done to detect if changes were made.
+  Default implementations are available for Kubernetes and protobuf objects, and will fallback to `reflect.DeepEqual`.
+* A `Name`, `Namespace`, `Labels`, and `LabelSelector` may optionally be included, for use with filters (see below).
+
+## Derived Collections
+
+The core of the framework is in the ability to derive collections from others.
+
+In general, these are built by providing some `func(inputs...) outputs...`.
+While more could be expressed, there are currently three forms implemented.
+
+* `func() *O` via `NewSingleton`
+  * This generates a collection that has a single value. An example would be some global configuration.
+* `func(input I) *O` via `NewCollection`
+  * This generates a one-to-one mapping of input to output. An example would be a transformation from a `Pod` type to a generic `Workload` type.
+* `func(input I) []O` via `NewManyCollection`
+  * This generates a one-to-many mapping of input to output. An example would be a transformation from a `Service` to a _set_ of `Endpoint` types.
+  * The order of the response does not matter. Each response must have a unique key.
+
+The form used and input type only represent the _primary dependencies_, indicating the cardinality.
+Each transformation can additionally include an arbitrary number of dependencies, fetching data from other collections.
+
+For example, a simple `Singleton` example that keeps track of the number of `ConfigMap`s in the cluster:
+
+```go
+ConfigMapCount := krt.NewSingleton[int](func(ctx krt.HandlerContext) *int {
+    cms := krt.Fetch(ctx, ConfigMaps)
+    return ptr.Of(len(cms))
+})
+```
+
+The `Fetch` operation enables querying against other collections.
+If the result of the `Fetch` operation changes, the collection will automatically be recomputed; the framework handles the state and event detection.
+In the above example, the provided function will be called (at least) every time there is a change to a configmap.
+The `ConfigMapCount` collection will produce events only when the count changes.
+The framework will automatically suppress events when nothing has changed.
+
+### Picking a collection type
+
+There are a variety of collection types available.
+Picking these is about simplicity, usability, and performance.
+
+The `NewSingleton` form (`func() *O`), in theory, could be used universally.
+Consider a transformation from `Pod` to `SimplePod`:
+
+```go
+SimplePods := krt.NewSingleton[SimplePod](func(ctx krt.HandlerContext) *[]SimplePod {
```

Comment:
Ah ok so this is effectively the "many-to-many" relationship I mentioned before; I can see why this is probably rarer and may not make as much sense.